File: typescript/examples/cancel_order.ts
import { V1Transaction } from "../src";
import { asPubkeyStr } from "../src";
import {
  cancelBodySigned,
  cancelAllBodySigned,
} from "../src/domain/order/client";
import { generateCancelAllSalt, OrderSide } from "../src/program";
import { scalePriceSize } from "../src/shared";
import {
  getKeypair,
  login,
  marketAndOrderbook,
  quoteDepositMint,
  restClient,
  runExample,
  unixTimestamp,
} from "./common";

async function main() {
  const client = restClient();
  const keypair = getKeypair();
  await login(client, keypair);
  const pubkey = keypair.publicKey.toBase58();

  const snapshot = await client.orders().getUserOrders(50);
  const limitOrder = snapshot.orders.find((o) => o.order_type === "limit");

  if (!limitOrder) {
    console.log("No open limit orders to cancel.");
    return;
  }

  const orderHash = limitOrder.order_hash;
  const orderbookId = limitOrder.orderbook_id;

  const cancel = cancelBodySigned(orderHash, asPubkeyStr(pubkey), keypair);
  const cancelled = await client.orders().cancel(cancel);
  console.log(`cancelled: ${cancelled.order_hash} remaining=${cancelled.remaining}`);

  const timestamp = unixTimestamp();
  const salt = generateCancelAllSalt();
  const cancelAll = cancelAllBodySigned(
    asPubkeyStr(pubkey),
    orderbookId,
    timestamp,
    salt,
    keypair
  );
  const cleared = await client.orders().cancelAll(cancelAll);
  console.log(`cancel-all removed ${cleared.count} order(s) in ${cleared.orderbook_id}`);

  // Cleanup: cancelling the order released its locked collateral back into
  // the global pool. Withdraw that amount to the user's token account so the
  // companion `submit_order` → `cancel_order` cycle is net-neutral on the
  // wallet's balance and the global pool.
  const [, orderbook] = await marketAndOrderbook(client);
  const rules = await client.orderbooks().decimals(orderbook.orderbookId);
  const orderQuoteAmount = scalePriceSize(
    rules.tradingRules.priceQuantum,
    "1",
    OrderSide.BID,
    rules
  ).quoteAtoms;
  const mint = quoteDepositMint(orderbook);
  const withdrawIx = client
    .positions()
    .withdrawFromGlobal()
    .user(keypair.publicKey)
    .mint(mint)
    .amount(orderQuoteAmount)
    .buildIx();
  const context = await client.transactionContext();
  const tx = V1Transaction.compile([withdrawIx], keypair.publicKey, context);
  const signed = tx.sign([keypair]);
  const sig = await client.rpc().submitSignedTransaction(signed);
  await client.rpc().confirmSignature(sig, context.lastValidBlockHeight);
  console.log(`withdraw_from_global: confirmed ${sig}`);
}

void runExample(main);

File: python/examples/cancel_order.py
"""Cancel a single order by hash, cancel all orders in an orderbook, and withdraw the released collateral back from the global pool."""

import asyncio

from common import (
    client as make_client,
)
from common import (
    get_keypair,
    login,
    market_and_orderbook,
    quote_deposit_mint,
    unix_timestamp,
)

from lightcone_sdk.domain.order import CancelAllBody, CancelBody
from lightcone_sdk.program.orders import (
    generate_cancel_all_salt,
    sign_cancel_all,
    sign_cancel_order,
)
from lightcone_sdk.program.types import OrderSide
from lightcone_sdk.shared.scaling import scale_price_size
from lightcone_sdk.shared.signing import SigningStrategy


async def main():
    client = make_client()
    keypair = get_keypair()
    client.set_signing_strategy(SigningStrategy.native(keypair))
    await login(client, keypair)
    pubkey = str(keypair.pubkey())

    # 1. Find an open limit order
    snapshot = await client.orders().get_user_orders(50)
    limit_order = next((o for o in snapshot.orders if o.order_type == "limit"), None)

    if limit_order is None:
        print("No open limit orders to cancel.")
        await client.close()
        return

    order_hash = limit_order.order_hash
    orderbook_id = limit_order.orderbook_id

    # 2. Cancel a single order
    signature = sign_cancel_order(order_hash, keypair)
    cancelled = await client.orders().cancel(
        CancelBody(order_hash=order_hash, maker=pubkey, signature=signature)
    )
    print(f"cancelled: {cancelled.order_hash} remaining={cancelled.remaining}")

    # 3. Cancel all orders in an orderbook
    timestamp = unix_timestamp()
    salt = generate_cancel_all_salt()
    cancel_all_sig = sign_cancel_all(pubkey, orderbook_id, timestamp, salt, keypair)
    cleared = await client.orders().cancel_all(
        CancelAllBody(
            user_pubkey=pubkey,
            orderbook_id=orderbook_id,
            signature=cancel_all_sig,
            timestamp=timestamp,
            salt=salt,
        )
    )
    print(f"cancel-all removed {cleared.count} order(s) in {cleared.orderbook_id}")

    # 4. Cleanup: cancelling the order released its locked collateral back into
    #    the global pool. Withdraw that amount to the user's token account so
    #    the companion submit_order → cancel_order cycle is net-neutral on the
    #    wallet's balance and the global pool.
    _, orderbook = await market_and_orderbook(client)
    rules = await client.orderbooks().decimals(orderbook.orderbook_id)
    order_quote_amount = scale_price_size(
        rules.trading_rules.price_quantum, "1", int(OrderSide.BID), rules
    ).quote_atoms
    mint = quote_deposit_mint(orderbook)
    withdraw_ix = (
        client.positions()
        .withdraw_from_global()
        .user(keypair.pubkey())
        .mint(mint)
        .amount(order_quote_amount)
        .build_ix()
    )
    withdraw_tx = await client.rpc().build_transaction(
        [withdraw_ix], keypair.pubkey(), await client.transaction_context()
    )
    withdraw_result = await client.sign_and_submit_tx_confirmed_with_slot(withdraw_tx)
    print(f"withdraw_from_global: confirmed {withdraw_result.signature}")

    await client.close()


asyncio.run(main())
